File: correctness/gen_hard_variants.R
#!/usr/bin/env Rscript
# Generate a synthetic corpus of PATHOLOGICAL / hard variants for the consequence
# classes we target — placed at *exact* offsets relative to transcript features
# (which random sampling almost never hits), to stress-test duckvep vs Ensembl VEP.
#
# Classes: splice-boundary indels that SPAN a donor/acceptor/5th-base/donor-region/
# polypyrimidine/splice-region boundary (the interval-vs-point case), CDS frameshift
# (1 bp) / inframe (3 bp) indels, and in-codon MNVs.
#
# Stack (no Python): DuckDB SQL for coordinates, samtools faidx for ref bases,
# vcfppR's writer for output — the same tooling as ~/duckhts.
#
# Usage: gen_hard_variants.R <exons.parquet> <fasta> <transcripts.parquet> <chrom> <out.vcf>
suppressMessages(library(vcfppR))
a <- commandArgs(trailingOnly = TRUE)
EXONS <- a[1]; FASTA <- a[2]; TX <- a[3]; CHROM <- a[4]; OUT <- a[5]
DUCKDB <- Sys.getenv("DUCKVEP_DUCKDB", ".tools/duckdb")

duck <- function(sql) {
  txt <- system2(DUCKDB, c("-csv", "-noheader", "-c", sql), stdout = TRUE)
  if (length(txt) == 0) return(data.frame())
  read.csv(text = paste(txt, collapse = "\n"), header = FALSE, stringsAsFactors = FALSE)
}

# Coding, multi-exon transcripts on this contig (both strands), MANE/canonical.
tx <- duck(sprintf("SELECT transcript_id FROM read_parquet('%s')
  WHERE chrom='%s' AND biotype='protein_coding' AND (mane_select OR canonical)
  ORDER BY gene_symbol LIMIT 40", TX, CHROM))[[1]]
ex <- duck(sprintf("SELECT transcript_id, start, \"end\" FROM read_parquet('%s')
  WHERE transcript_id IN ('%s') ORDER BY transcript_id, start",
  EXONS, paste(tx, collapse = "','")))
names(ex) <- c("tid", "start", "end")

# Introns = gaps between position-sorted exons; keep windows around each boundary.
introns <- do.call(rbind, lapply(split(ex, ex$tid), function(d) {
  d <- d[order(d$start), ]
  if (nrow(d) < 2) return(NULL)
  data.frame(istart = head(d$end, -1) + 1, iend = tail(d$start, -1) - 1)
}))
introns <- introns[introns$iend - introns$istart >= 40, ]

# One samtools faidx call for all donor/acceptor windows; index by "chr:a-b".
regions <- unique(c(sprintf("%s:%d-%d", CHROM, introns$istart - 6, introns$istart + 8),
                    sprintf("%s:%d-%d", CHROM, introns$iend - 18, introns$iend + 8)))
fa <- system2("samtools", c("faidx", FASTA, regions), stdout = TRUE)
h <- grepl("^>", fa); ids <- sub("^>", "", fa[h])
seqs <- toupper(tapply(fa[!h], cumsum(h)[!h], paste, collapse = ""))
win <- setNames(as.character(seqs), ids)
# Genomic [a,b] inclusive from whichever window contains it (1-based slice).
getseq <- function(a, b) {
  for (nm in names(win)) {
    p <- as.integer(strsplit(sub(".*:", "", nm), "-")[[1]])
    if (a >= p[1] && b <= p[2]) return(substr(win[[nm]], a - p[1] + 1, b - p[2] + nchar(win[[nm]])))
  }
  NA_character_
}

recs <- list(); seen <- new.env()
add <- function(pos, ref, alt, tag) {
  if (is.na(ref) || is.na(alt) || pos < 2 || grepl("N", ref) || grepl("N", alt)) return()
  k <- paste(pos, ref, alt); if (!is.null(seen[[k]])) return(); seen[[k]] <- TRUE
  recs[[length(recs) + 1]] <<- sprintf("%s\t%d\t%s\t%s\t%s\t.\t.\t.", CHROM, pos, tag, ref, alt)
}
del <- function(s, e, tag) add(s - 1, getseq(s - 1, e), getseq(s - 1, s - 1), tag)  # VCF anchor-base del
ins <- function(p, b, tag) { a <- getseq(p, p); add(p, a, paste0(a, b), tag) }
comp <- c(A = "T", T = "A", C = "G", G = "C")

for (i in seq_len(nrow(introns))) {
  is <- introns$istart[i]; ie <- introns$iend[i]
  del(is - 1, is + 1, "del_span_donor5p")      # exon|GT (spans 5' boundary)
  del(is + 3, is + 5, "del_donor_region")
  del(is + 4, is + 4, "del_5th_base")
  del(ie - 1, ie + 1, "del_span_acceptor3p")   # AG|exon (spans 3' boundary)
  del(ie - 12, ie - 8, "del_polypyrimidine")
  del(is + 2, is + 7, "del_splice_region")
  ins(is + 1, "AT", "ins_donor")
  ins(ie - 3, "GC", "ins_acceptor_region")
  del(is - 1, is - 1, "del_fs1_exonend")        # 1 bp -> frameshift
  del(is - 4, is - 2, "del_inframe3_exonend")   # 3 bp -> inframe
  r <- getseq(ie + 6, ie + 8)                   # in-codon MNV in next exon
  if (!is.na(r) && nchar(r) == 3) add(ie + 6, r, paste(comp[strsplit(r, "")[[1]]], collapse = ""), "mnv3_exon")
}

w <- methods::new(vcfppR::vcfwriter, OUT, "VCFv4.2")
w$addLine("##source=correctness/gen_hard_variants.R")
w$addContig(CHROM)
ord <- order(as.integer(sub(".*\t(\\d+)\t.*", "\\1", sub("^[^\t]*\t", "", recs))))
for (r in recs[ord]) w$writeline(r)
w$close()
cat(sprintf("generated %d hard variants across %d transcripts\n", length(recs), length(tx)), file = stderr())

File: tools/src/gen_hard_variants.rs
//! gen-hard-variants — generate a synthetic corpus of PATHOLOGICAL hard variants
//! for the consequence classes we target, placed at *exact* offsets relative to
//! transcript features (which random sampling almost never hits). Pure Rust:
//! parquet for coordinates, `fastvep_cache::FastaReader` (noodles-backed) for ref
//! bases — no bash/R/vcfppR/Python.
//!
//! Usage: gen-hard-variants <exons.parquet> <transcripts.parquet> <fasta> <chrom> <out.vcf>

use anyhow::{Context, Result};
use arrow::array::{Array, BooleanArray, Int64Array, RecordBatch, StringArray};
use fastvep_cache::fasta::FastaReader;
use parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder;
use std::collections::{BTreeSet, HashMap, HashSet};
use std::fs::File;
use std::io::Write;

type Rec = (i64, String, String, String); // (pos, id/tag, ref, alt)

fn cstr<'a>(b: &'a RecordBatch, n: &str) -> &'a StringArray {
    b.column(b.schema().index_of(n).unwrap())
        .as_any()
        .downcast_ref()
        .unwrap()
}
fn ci64<'a>(b: &'a RecordBatch, n: &str) -> &'a Int64Array {
    b.column(b.schema().index_of(n).unwrap())
        .as_any()
        .downcast_ref()
        .unwrap()
}
fn cbool<'a>(b: &'a RecordBatch, n: &str) -> &'a BooleanArray {
    b.column(b.schema().index_of(n).unwrap())
        .as_any()
        .downcast_ref()
        .unwrap()
}

fn seq(fa: &FastaReader, chrom: &str, s: i64, e: i64) -> Option<String> {
    if s < 1 {
        return None;
    }
    let v = fa.fetch(chrom, s as u64, e as u64).ok()?;
    let s = String::from_utf8_lossy(&v).into_owned();
    if s.is_empty() || s.contains('N') {
        None
    } else {
        Some(s)
    }
}

/// VCF anchor-base deletion of genomic [s, e] inclusive (ref = anchor..e, alt = anchor).
fn del(fa: &FastaReader, chrom: &str, r: &mut BTreeSet<Rec>, s: i64, e: i64, tag: &str) {
    if s - 1 < 2 {
        return;
    }
    if let (Some(rf), Some(an)) = (seq(fa, chrom, s - 1, e), seq(fa, chrom, s - 1, s - 1)) {
        r.insert((s - 1, tag.into(), rf, an));
    }
}
/// VCF insertion of `b` after position p (ref = base@p, alt = base@p + b).
fn ins(fa: &FastaReader, chrom: &str, r: &mut BTreeSet<Rec>, p: i64, b: &str, tag: &str) {
    if let Some(an) = seq(fa, chrom, p, p) {
        r.insert((p, tag.into(), an.clone(), format!("{an}{b}")));
    }
}

fn main() -> Result<()> {
    let a: Vec<String> = std::env::args().skip(1).collect();
    if a.len() < 5 {
        anyhow::bail!("usage: gen-hard-variants <exons.parquet> <transcripts.parquet> <fasta> <chrom> <out.vcf>");
    }
    let (exons_p, tx_p, fasta_p, chrom, out_p) = (&a[0], &a[1], &a[2], a[3].clone(), &a[4]);

    // Coding, MANE/canonical transcripts on this contig.
    let mut keep: HashSet<String> = HashSet::new();
    for batch in ParquetRecordBatchReaderBuilder::try_new(File::open(tx_p)?)?.build()? {
        let b = batch?;
        let (tid, ch, bt) = (
            cstr(&b, "transcript_id"),
            cstr(&b, "chrom"),
            cstr(&b, "biotype"),
        );
        let (mane, canon) = (cbool(&b, "mane_select"), cbool(&b, "canonical"));
        for i in 0..b.num_rows() {
            if ch.value(i) == chrom
                && bt.value(i) == "protein_coding"
                && (mane.value(i) || canon.value(i))
            {
                keep.insert(tid.value(i).to_string());
            }
        }
    }

    // Exons (genomic) per kept transcript.
    let mut exons: HashMap<String, Vec<(i64, i64)>> = HashMap::new();
    for batch in ParquetRecordBatchReaderBuilder::try_new(File::open(exons_p)?)?.build()? {
        let b = batch?;
        let (tid, s, e) = (
            cstr(&b, "transcript_id"),
            ci64(&b, "start"),
            ci64(&b, "end"),
        );
        for i in 0..b.num_rows() {
            let t = tid.value(i);
            if keep.contains(t) {
                exons
                    .entry(t.to_string())
                    .or_default()
                    .push((s.value(i), e.value(i)));
            }
        }
    }

    let fa = FastaReader::from_reader(File::open(fasta_p).context("open fasta")?)?;
    let mut recs: BTreeSet<Rec> = BTreeSet::new();
    for ex in exons.values() {
        let mut ex = ex.clone();
        ex.sort();
        for w in ex.windows(2) {
            let (istart, iend) = (w[0].1 + 1, w[1].0 - 1); // intron (genomic)
            if iend - istart < 40 {
                continue;
            }
            del(
                &fa,
                &chrom,
                &mut recs,
                istart - 1,
                istart + 1,
                "del_span_donor5p",
            );
            del(
                &fa,
                &chrom,
                &mut recs,
                istart + 3,
                istart + 5,
                "del_donor_region",
            );
            del(
                &fa,
                &chrom,
                &mut recs,
                istart + 4,
                istart + 4,
                "del_5th_base",
            );
            del(
                &fa,
                &chrom,
                &mut recs,
                iend - 1,
                iend + 1,
                "del_span_acceptor3p",
            );
            del(
                &fa,
                &chrom,
                &mut recs,
                iend - 12,
                iend - 8,
                "del_polypyrimidine",
            );
            del(
                &fa,
                &chrom,
                &mut recs,
                istart + 2,
                istart + 7,
                "del_splice_region",
            );
            ins(&fa, &chrom, &mut recs, istart + 1, "AT", "ins_donor");
            ins(
                &fa,
                &chrom,
                &mut recs,
                iend - 3,
                "GC",
                "ins_acceptor_region",
            );
            del(
                &fa,
                &chrom,
                &mut recs,
                istart - 1,
                istart - 1,
                "del_fs1_exonend",
            );
            del(
                &fa,
                &chrom,
                &mut recs,
                istart - 4,
                istart - 2,
                "del_inframe3_exonend",
            );
            if let Some(r) = seq(&fa, &chrom, iend + 6, iend + 8) {
                if r.len() == 3 {
                    let comp: String = r
                        .chars()
                        .map(|c| match c {
                            'A' => 'C',
                            'C' => 'G',
                            'G' => 'T',
                            'T' => 'A',
                            o => o,
                        })
                        .collect();
                    recs.insert((iend + 6, "mnv3_exon".into(), r, comp));
                }
            }
        }
    }

    let mut out = File::create(out_p)?;
    writeln!(
        out,
        "##fileformat=VCFv4.2\n##source=tools/gen-hard-variants\n##contig=<ID={chrom}>"
    )?;
    writeln!(out, "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO")?;
    for (pos, tag, r, alt) in &recs {
        writeln!(out, "{chrom}\t{pos}\t{tag}\t{r}\t{alt}\t.\t.\t.")?;
    }
    eprintln!(
        "generated {} hard variants across {} transcripts",
        recs.len(),
        exons.len()
    );
    Ok(())
}
